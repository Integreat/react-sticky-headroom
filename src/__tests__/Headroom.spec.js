// @flow

import * as React from 'react'
import { mount, shallow } from 'enzyme'
import Headroom from '../Headroom'
import 'jest-styled-components'

const pinStart = 10
const height = 100
const scrollHeight = 50

describe('Headroom', () => {
  const MockNode = () => <div />

  const createComponent = props => mount(
    <Headroom scrollHeight={50}
              height={100}
              {...props}>
      <MockNode />
    </Headroom>)

  it('should have correct default state', () => {
    const component = createComponent()
    expect(component.state()).toEqual({ mode: 'static', transition: 'none', keyframes: null })
    expect(component.prop('pinStart')).toEqual(0)
  })

  it('should render with values from its state', () => {
    const component = createComponent()
    component.setState({ transform: 42, stickyTop: 24 })
    expect(component).toMatchSnapshot()
  })

  it('should render with no stickyAncestor supplied', () => {
    const component = shallow(
      <Headroom scrollHeight={50}><MockNode /></Headroom>)
    expect(component).toMatchSnapshot()
  })

  it('should attach and detach listener for onscroll event', () => {
    const originalAdd = window.addEventListener
    const originalRemove = window.removeEventListener
    window.addEventListener = jest.fn()
    window.removeEventListener = jest.fn()

    const component = mount(<Headroom scrollHeight={50}><MockNode /></Headroom>)
    const handleEventCallback = component.instance().handleEvent
    expect(window.addEventListener).toHaveBeenCalledWith('scroll', handleEventCallback)
    component.unmount()
    expect(window.removeEventListener).toHaveBeenCalledWith('scroll', handleEventCallback)

    window.addEventListener = originalAdd
    window.removeEventListener = originalRemove
  })

  it('should request animation frame and update, on handleEvent', () => {
    const originalRaf = window.requestAnimationFrame
    window.requestAnimationFrame = jest.fn()
    const component = mount(<Headroom scrollHeight={50}><MockNode /></Headroom>)
    component.instance().update = jest.fn()

    // Call first time
    component.instance().handleEvent()
    expect(window.requestAnimationFrame).toHaveBeenCalledTimes(1)
    expect(component.instance().update).toHaveBeenCalledTimes(0)
    // Now perform the raf
    window.requestAnimationFrame.mock.calls[0][0]()
    expect(component.instance().update).toHaveBeenCalledTimes(1)

    window.requestAnimationFrame = originalRaf
  })

  describe('update', () => {
    it('should set correct state, if user hasn\'t scrolled beyond pinStart', () => {
      const component = createComponent({ pinStart, height, scrollHeight })
      const scrollTo = scrollTo => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }
      scrollTo(0)
      scrollTo(pinStart / 2)
      expect(component.state()).toEqual({ mode: 'static', transition: 'none', keyframes: null })
    })

    it('should set correct state, if user has scrolled down to pinStart + scrollHeight/2', () => {
      const component = createComponent({ pinStart, height, scrollHeight })
      const scrollTo = scrollTo => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }

      scrollTo(0)
      scrollTo(pinStart + scrollHeight / 2)
      expect(component.state()).toEqual({ mode: 'static', transition: 'none', keyframes: null })
    })

    it('should set correct state, if user has scrolled down and back up again', () => {
      const component = createComponent({ pinStart, height, scrollHeight, keyframes: null })
      const scrollTo = scrollTo => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }

      scrollTo(pinStart)
      expect(component.state()).toEqual({ mode: 'static', transition: 'none', keyframes: null })

      const offset = 5
      scrollTo(pinStart + scrollHeight)
      // Header is completely transformed to the top
      expect(component.state()).toEqual({ mode: 'static', transition: 'none', keyframes: null })

      scrollTo(pinStart + scrollHeight + offset)
      // Header should be unpinned now, transitions should be off though
      expect(component.state()).toEqual({ mode: 'unpinned', transition: 'none', keyframes: null })

      scrollTo(pinStart + scrollHeight)
      // Header should be pinned with transition, because we're scrolling upwards
      expect(component.state()).toEqual({ mode: 'pinned', transition: 'normal', keyframes: null })

      scrollTo(pinStart)
      expect(component.state()).toEqual({ mode: 'static', transition: 'pinned-to-static', keyframes: expect.any(Object) })
    })

    it('shouldn\'t update state update is called with same scrollTop', () => {
      const component = createComponent({ pinStart, height, scrollHeight })
      const scrollTo = scrollTo => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }
      const offset = 5
      scrollTo(pinStart + scrollHeight + offset)
      expect(component.state()).toEqual({ mode: 'unpinned', transition: 'none', keyframes: null })
      scrollTo(pinStart + scrollHeight + offset)
      expect(component.state()).toEqual({ mode: 'unpinned', transition: 'none', keyframes: null })
    })

    it('should call onStickyTopChanged if mode has changed', () => {
      const onStickyTopChanged = jest.fn()
      const component = createComponent({ pinStart, height, scrollHeight, onStickyTopChanged })
      const scrollTo = scrollTo => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }

      scrollTo(pinStart + scrollHeight + 10)
      expect(onStickyTopChanged).toHaveBeenCalledWith(scrollHeight)
      scrollTo(0)
      expect(onStickyTopChanged).toHaveBeenCalledWith(height - scrollHeight)
    })
  })

  it('should render correct if state is static, no transition', () => {
    const component = createComponent({ pinStart, height, scrollHeight })
    component.setState({ mode: 'static', transition: 'none' })
    expect(component).toMatchSnapshot()
  })

  it('should render correct if state is unpinned, no transition', () => {
    const component = createComponent({ pinStart, height, scrollHeight })
    component.setState({ mode: 'unpinned', transition: 'none' })
    expect(component).toMatchSnapshot()
  })

  it('should render correct if state is unpinned, transition', () => {
    const component = createComponent({ pinStart, height, scrollHeight })
    component.setState({ mode: 'unpinned', transition: 'normal' })
    expect(component).toMatchSnapshot()
  })

  it('should render correct if state is pinned, transition', () => {
    const component = createComponent({ pinStart, height, scrollHeight })
    component.setState({ mode: 'pinned', transition: 'normal' })
    expect(component).toMatchSnapshot()
  })
})
