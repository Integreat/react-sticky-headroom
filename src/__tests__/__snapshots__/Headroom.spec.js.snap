// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headroom should render correct if state is pinned, transition 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: 0px;
  z-index: 1;
  -webkit-transform: translateY(0px);
  -ms-transform: translateY(0px);
  transform: translateY(0px);
  -webkit-animation-duration: 0.2s;
  animation-duration: 0.2s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
  -webkit-transition: -webkit-transform 0.2s ease-out;
  -webkit-transition: transform 0.2s ease-out;
  transition: transform 0.2s ease-out;
}

<Headroom
  height={100}
  pinStart={10}
  scrollHeight={50}
>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={false}
    top={0}
    transition="normal"
    translateY={0}
  >
    <StyledComponent
      animateUpFrom={null}
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "position:",
              [Function],
              ";top:",
              [Function],
              "px;z-index:1;transform:translateY(",
              [Function],
              "px);animation-duration:0.2s;animation-timing-function:ease-out;",
              [Function],
              " ",
              [Function],
              " ",
              [Function],
            ],
          },
          "displayName": "Headroom__HeaderWrapper",
          "render": [Function],
          "styledComponentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      static={false}
      top={0}
      transition="normal"
      translateY={0}
    >
      <div
        className="c0"
      >
        <MockNode>
          <div />
        </MockNode>
      </div>
    </StyledComponent>
  </Headroom__HeaderWrapper>
</Headroom>
`;

exports[`Headroom should render correct if state is static, no transition 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: -50px;
  z-index: 1;
  -webkit-transform: translateY(0px);
  -ms-transform: translateY(0px);
  transform: translateY(0px);
  -webkit-animation-duration: 0.2s;
  animation-duration: 0.2s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
  -webkit-transition: none;
  transition: none;
}

<Headroom
  height={100}
  pinStart={10}
  scrollHeight={50}
>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={true}
    top={-50}
    transition="none"
    translateY={0}
  >
    <StyledComponent
      animateUpFrom={null}
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "position:",
              [Function],
              ";top:",
              [Function],
              "px;z-index:1;transform:translateY(",
              [Function],
              "px);animation-duration:0.2s;animation-timing-function:ease-out;",
              [Function],
              " ",
              [Function],
              " ",
              [Function],
            ],
          },
          "displayName": "Headroom__HeaderWrapper",
          "render": [Function],
          "styledComponentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      static={true}
      top={-50}
      transition="none"
      translateY={0}
    >
      <div
        className="c0"
      >
        <MockNode>
          <div />
        </MockNode>
      </div>
    </StyledComponent>
  </Headroom__HeaderWrapper>
</Headroom>
`;

exports[`Headroom should render correct if state is unpinned, no transition 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: 0px;
  z-index: 1;
  -webkit-transform: translateY(-50px);
  -ms-transform: translateY(-50px);
  transform: translateY(-50px);
  -webkit-animation-duration: 0.2s;
  animation-duration: 0.2s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
}

<Headroom
  height={100}
  pinStart={10}
  scrollHeight={50}
>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={false}
    top={0}
    transition="none"
    translateY={-50}
  >
    <StyledComponent
      animateUpFrom={null}
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "position:",
              [Function],
              ";top:",
              [Function],
              "px;z-index:1;transform:translateY(",
              [Function],
              "px);animation-duration:0.2s;animation-timing-function:ease-out;",
              [Function],
              " ",
              [Function],
              " ",
              [Function],
            ],
          },
          "displayName": "Headroom__HeaderWrapper",
          "render": [Function],
          "styledComponentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      static={false}
      top={0}
      transition="none"
      translateY={-50}
    >
      <div
        className="c0"
      >
        <MockNode>
          <div />
        </MockNode>
      </div>
    </StyledComponent>
  </Headroom__HeaderWrapper>
</Headroom>
`;

exports[`Headroom should render correct if state is unpinned, transition 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: 0px;
  z-index: 1;
  -webkit-transform: translateY(-50px);
  -ms-transform: translateY(-50px);
  transform: translateY(-50px);
  -webkit-animation-duration: 0.2s;
  animation-duration: 0.2s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
  -webkit-transition: -webkit-transform 0.2s ease-out;
  -webkit-transition: transform 0.2s ease-out;
  transition: transform 0.2s ease-out;
}

<Headroom
  height={100}
  pinStart={10}
  scrollHeight={50}
>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={false}
    top={0}
    transition="normal"
    translateY={-50}
  >
    <StyledComponent
      animateUpFrom={null}
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "position:",
              [Function],
              ";top:",
              [Function],
              "px;z-index:1;transform:translateY(",
              [Function],
              "px);animation-duration:0.2s;animation-timing-function:ease-out;",
              [Function],
              " ",
              [Function],
              " ",
              [Function],
            ],
          },
          "displayName": "Headroom__HeaderWrapper",
          "render": [Function],
          "styledComponentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      static={false}
      top={0}
      transition="normal"
      translateY={-50}
    >
      <div
        className="c0"
      >
        <MockNode>
          <div />
        </MockNode>
      </div>
    </StyledComponent>
  </Headroom__HeaderWrapper>
</Headroom>
`;

exports[`Headroom should render with no stickyAncestor supplied 1`] = `
<Fragment>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={true}
    top={-50}
    transition="none"
    translateY={0}
  >
    <MockNode />
  </Headroom__HeaderWrapper>
</Fragment>
`;

exports[`Headroom should render with values from its state 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: -50px;
  z-index: 1;
  -webkit-transform: translateY(0px);
  -ms-transform: translateY(0px);
  transform: translateY(0px);
  -webkit-animation-duration: 0.2s;
  animation-duration: 0.2s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
  -webkit-transition: none;
  transition: none;
}

<Headroom
  height={100}
  pinStart={0}
  scrollHeight={50}
>
  <Headroom__HeaderWrapper
    animateUpFrom={null}
    static={true}
    top={-50}
    transition="none"
    translateY={0}
  >
    <StyledComponent
      animateUpFrom={null}
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "position:",
              [Function],
              ";top:",
              [Function],
              "px;z-index:1;transform:translateY(",
              [Function],
              "px);animation-duration:0.2s;animation-timing-function:ease-out;",
              [Function],
              " ",
              [Function],
              " ",
              [Function],
            ],
          },
          "displayName": "Headroom__HeaderWrapper",
          "render": [Function],
          "styledComponentId": "Headroom__HeaderWrapper-sc-16d3zk-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      static={true}
      top={-50}
      transition="none"
      translateY={0}
    >
      <div
        className="c0"
      >
        <MockNode>
          <div />
        </MockNode>
      </div>
    </StyledComponent>
  </Headroom__HeaderWrapper>
</Headroom>
`;
